Use ES6 classes

Zielak opened the pull request:
I couldn't wait so I tried it myself.
I only changed how classes are defined. All build and testing tools are not changed.
#3

Commit:
move to ES6 classes

Commit:
one fix after gulp

danielepolencic commented:
This looks great!

Let me review it and I'll soon merge it in!

💪 💪 💪

danielepolencic commented on src/actions/Wait.js:
```diff
@@ -78,4 +56,25 @@ export default Class(Action, {
+ * @property {String} parameters
+ * @readonly
+ **/
+Wait.prototype.parameters = {'milliseconds': 0};
```
No need for extra single quotes: `{milliseconds: 0}`

danielepolencic commented on src/core/Action.js:
```diff
@@ -20,21 +19,21 @@ import {ACTION} from '../constants';
-    BaseNode.prototype.initialize.call(this);
+  constructor() {
+    super();
   }
```
If I'm not mistaken, this can be omitted because the constructor for the parent class is called anyway.

Zielak replied:
Correct, good catch :)

danielepolencic commented on src/core/BaseNode.js:
```diff
@@ -245,5 +178,72 @@ export default Class(null, {
+ * @property {string} id
+ * @readonly
+ **/
+BaseNode.prototype.id = null;
```
Unfortunately this code contains a very subtle bug. Because the prototype is shared amongst all instances, if you change the `id` for one of them, all of the other instances will change too. More on this: https://stackoverflow.com/questions/16911132/how-are-properties-shared-across-instances-using-javascript-prototypes

Consider moving this as an instance variable:

```js
class B {
  constructor() {
    this.id = null;
  }
}
```

Ironically, moving all the properties to be instance properties will make the code much terse.

Zielak replied:
Hmm, but we're never changing or accessing that `BaseNode.prototype.id` value directly, or did I miss something?
When creating new instance of BaseNode (or anything that extends from it), its constructor creates new property `id` for that instance. So after new instance has been created, calling `newNode.id` will give you that instance's id, instead of the one from the prototype.

danielepolencic replied:
Good point. In that case I think we can drop it.

We have the same properties declared twice: https://github.com/Zielak/behavior3js/blob/35ed667ab0ea41c112d00e464aa783005f35b672/src/core/BaseNode.js#L29-L38 and https://github.com/Zielak/behavior3js/blob/35ed667ab0ea41c112d00e464aa783005f35b672/src/core/BaseNode.js#L184-L249

What do you think?

Zielak replied:
Yup, `id` could be dropped from this prototype. What about other properties?
For example: `name` and `title` are re-defined in `Wait.prototype`, but `Succeeder` only defines `name`.
Seems like some properties of prototypes are not required.
Now that I think about it, maybe it's better to keep other properties of BaseNode's prototype as they are. They serve a nice description for what can be used when extending each class.

danielepolencic commented on src/decorators/RepeatUntilFailure.js:
```diff
@@ -89,4 +67,25 @@ export default Class(Decorator, {
+ * @property {String} parameters
+ * @readonly
+ **/
+RepeatUntilFailure.prototype.parameters = {'maxLoop': -1}
```
no quotes needed

danielepolencic commented on src/decorators/MaxTime.js:
```diff
@@ -90,4 +68,25 @@ export default Class(Decorator, {
+ * @property {String} parameters
+ * @readonly
+ **/
+MaxTime.prototype.parameters = {'maxTime': 0};
```
no quotes needed

danielepolencic commented on src/decorators/Limiter.js:
```diff
@@ -91,4 +69,25 @@ export default Class(Decorator, {
+ * @property {String} parameters
+ * @readonly
+ **/
+Limiter.prototype.parameters = {'maxLoop': 1};
```
no quotes needed

danielepolencic commented on src/decorators/RepeatUntilSuccess.js:
```diff
@@ -89,4 +67,25 @@ export default Class(Decorator, {
+ * @property {String} parameters
+ * @readonly
+ **/
+RepeatUntilSuccess.prototype.parameters = {'maxLoop': -1};
```
no quotes needed

danielepolencic commented on src/decorators/Repeater.js:
```diff
@@ -88,4 +66,25 @@ export default Class(Decorator, {
+ * @property {String} parameters
+ * @readonly
+ **/
+Repeater.prototype.parameters = {'maxLoop': -1};
```
no quotes needed

danielepolencic commented on test/core/BehaviorTree-Serialization.js:
```diff
@@ -111,7 +110,7 @@ suite('Core: Behavior Tree - Serialization', function() {
     test('Load JSON model with custom nodes', function() {
         var tree = new BehaviorTree();
-        var CustomNode = Class(Condition);
+        var CustomNode = Condition;
```
can we remove this line and replace all the occurrences of `CustomNode` with `Condition`?

danielepolencic commented on test/core/BehaviorTree-Serialization.js:
```diff
@@ -149,7 +148,7 @@ suite('Core: Behavior Tree - Serialization', function() {
     test('Dump JSON model', function() {
         var tree = new BehaviorTree();
-        var CustomNode = Class(Condition);
+        var CustomNode = Condition;
```
same as above. Perhaps we can remove this line and replace all the occurrences of `CustomNode` with `Condition`

danielepolencic reviewed (changes requested):
Really great stuff! There's a small issue on how you used the prototype when you converted from ES5 to ES6. But a part from that this looks great.
Do you want to try and fix one class and remove the properties from the prototype? It's easier if we start with one, so that we can discuss approaches.

Commit:
changes after CR

Zielak commented:
Thanks for CR, updated :)

Commit:
remove `id` from the prototype

danielepolencic commented:
I'd like to merge this PR, but I think there're still cases where we are bitten by having properties on the prototype. To explain what I mean, have a look a the following example:

```js
function myClass(){}
myClass.prototype = {
    settings: {name: 'random'}
};

m1 = new myClass();
m2 = new myClass();

console.log(m1.settings.name) // 'random'
console.log(m2.settings.name) // 'random'

m1.settings.name = 'Dan'

console.log(m1.settings.name) // 'Dan'
console.log(m2.settings.name) // 'Dan' <- You might have expected 'random'
```

We're doing something similar here: https://github.com/behavior3/behavior3js/pull/30/files#diff-1cfa0cb7bc40f0bcf44081f319b96a43R127 . The array is shared by reference. The following line is similar.

I understand that having a description of the property is convenient, but perhaps we could use something like this? https://github.com/jsdoc3/jsdoc/issues/819

I'd like to avoid promoting practices that are very hard to debug and have an odd behaviour. Properties on the prototype is one of them.

What do you think?

Zielak commented:
I had to spend a minute on that example and now I see it. I think I've just gotten used to the current code with the prototype properties and their descriptions.
I'll get rid of those nightmares and move descriptions to constructor. Sadly, I don't see that jsdoc supports that kind of naming with `#`
```
/**
 * The actor's name.
 * @member Actor#name
 */
this.name = "leo";
```
but hopefully it'll change in the future.

Zielak commented:
Also, this change will require to change some tests:
```
test('Prototype', function() {
    assert.equal(Error.prototype.name, 'Error');
});
```

Zielak commented:
Another thing just popped up to my mind, at least for some properties:
```
class Error extends Action {
  // ...
  /**
   * Node name. Default to `Error`.
   * 
   * @readonly
   * @memberof Error
   */
  get name(){
    return 'Error';
  }
}
```

`category` will get inherited and `name` will get overriden when extending:

```
class Action {
	get name(){
		return 'Action';
	}
	get category(){
		return 'ACTIONCATEGORY';
	}
}

class Error extends Action {
	get name(){
		return 'Error';
	}
}

var e = new Error();
`e.name: ${e.name}, e.category: ${e.category}`
=> "e.name: Error, e.category: ACTIONCATEGORY"
```

Commit:
example of readonly properties definition

Zielak commented:
With my last commit I changed only actions for now (time to sleep).
Take a look at Wait.js, it has name and title but also parameters, which I renamed to properties, as `parameters` are deprecated (found that note on BaseNode.js). I'm not sure if we should keep `properties` read-only as I don't know the real usage of it. What do you think?

danielepolencic commented:
I think name, categories, etc. are used by the visual editor.
What if we slightly tweak the code like this: take `Action` as an example. The new action is:

```js
export default class Action extends BaseNode {
  constructor({name}) {
    this.category = ACTION;
    this.name = name;
  }
};
```

We could have `Error` as:

```js
export default class Error extends Action {
  constructor() {
     super({name: 'Error'});
  }
  tick() {/* ... */ }
}
```

I'm not worried about JSDoc, tbh.

Commit:
forget prototype properties, move descriptions to constructors, change name/title/etc inheritance

Commit:
some fixes for the tests

Commit:
fixed limiter and inverter

Commit:
"stuff" and default params for Wait

Commit:
comments + 2 tests

Commit:
wait, constructor 1 more assert

Zielak commented:
Hey, my changes are ready.

This makes it so `milliseconds` has default value of 0 if not present, and it doesn't break when there's nothing in constructor's arguments.
```
constructor({milliseconds = 0} = {}) {
```

Happy weekend!

danielepolencic commented:
Great effort 💪
